iotdb-core/consensus/src/test/java/org/apache/iotdb/consensus/ratis/RecoverReadTest.java

```diff
@@ -0,0 +1,176 @@
+                    .setRead(
+                        RatisConfig.Read.newBuilder()
+                            .setReadTimeout(TimeDuration.valueOf(30, TimeUnit.SECONDS))
+                            .build())
+                    .build())
+            .create();
+    miniCluster.start();
+  }
+
+  @After
+  public void tearUp() throws Exception {
+    miniCluster.cleanUp();
+  }
+
+  /* mimics the situation before this patch */
+  @Test
+  public void inconsistentReadAfterRestart() throws Exception {
+    final ConsensusGroupId gid = miniCluster.getGid();
+    final List<Peer> members = miniCluster.getPeers();
+
+    miniCluster.getServers().forEach(s -> s.createPeer(gid, members));
+
+    // first write 5 ops
+    TestUtils.write(miniCluster.getServer(0), gid, 10);
+    Assert.assertEquals(10, TestUtils.read(miniCluster.getServer(0), gid));
+
+    // set apply stall time
+    miniCluster.getStateMachines().stream()
+        .map(SlowRecoverStateMachine.class::cast)
+        .forEach(SlowRecoverStateMachine::setStall);
+
+    // shutdown and restart the cluster
+    miniCluster.restart();
+
+    // manually set the canServe flag to true to mimic original implementation
+    miniCluster.getServer(0).allowStaleRead(gid);
+
+    Assert.assertNotEquals(10, TestUtils.read(miniCluster.getServer(0), gid));
+  }
+
+  @Test
+  public void consistentRead() throws Exception {
+    final ConsensusGroupId gid = miniCluster.getGid();
+    final List<Peer> members = miniCluster.getPeers();
+
+    miniCluster.getServers().forEach(s -> s.createPeer(gid, members));
+
+    // first write 5 ops
+    TestUtils.write(miniCluster.getServer(0), gid, 10);
+    Assert.assertEquals(10, TestUtils.read(miniCluster.getServer(0), gid));
+
+    miniCluster.getStateMachines().stream()
+        .map(SlowRecoverStateMachine.class::cast)
+        .forEach(SlowRecoverStateMachine::setStall);
+
+    // shutdown and restart the cluster
+    miniCluster.restart();
+
+    // wait an active leader to serve linearizable read requests
+    miniCluster.waitUntilActiveLeader();
+
+    Assert.assertEquals(10, TestUtils.read(miniCluster.getServer(0), gid));
+  }
+
+  @Test
+  public void consistentReadTimeout() throws Exception {
+    final ConsensusGroupId gid = miniCluster.getGid();
+    final List<Peer> members = miniCluster.getPeers();
+
+    miniCluster.getServers().forEach(s -> s.createPeer(gid, members));
+
+    // first write 5 ops
+    TestUtils.write(miniCluster.getServer(0), gid, 10);
+    Assert.assertEquals(10, TestUtils.read(miniCluster.getServer(0), gid));
+
+    miniCluster.getStateMachines().stream()
+        .map(SlowRecoverStateMachine.class::cast)
+        .forEach(SlowRecoverStateMachine::setStall);
+
+    // shutdown and restart the cluster
+    miniCluster.restart();
+
+    // query during redo: get exception that ratis is under recovery
+    final ConsensusReadResponse readResponse = TestUtils.doRead(miniCluster.getServer(0), gid);
```

Comment:
The execution time for 10 logs is 5s, and the query timeout is 30s. Why is there an exception?

Reply:
It is caused by no-leader. No active leader will result in immediate return of failure.